### PR TITLE
Fix dynamic one-shot not working when no mcm and multiple output

### PR DESCRIPTION
**Pull request by mehrdad2m**
**Context:**
Specifying mcm_method="one-shot" when no MCMs are present raises an error when the return type is an iterable.

**Description of the Change:**
The crash happens in parse_native_mid_circuit_measurements which is a pennylane method and since in pennylane we exit the dynamic one-shot when no mcm is present, we are not expecting a no mcm situation which causes the crash. In this PR, we simply bypass the parse_native_mid_circuit_measurements function since there is no need to process mcm runs and only process the non mcm results
 
**Benefits:**
Dynamic-one-shot can work fine when no mcm is present and having multiple output.

**Related GitHub Issues:**
[sc-68752]

### CODE REVIEW

**Comment by mehrdad2m** [comment]
Note that I also added changelog for this PR (https://github.com/PennyLaneAI/catalyst/pull/1041) which was forgotten when it was merged.

**Review by dime10 — APPROVED** [review]
Nice 🥳

**Comment by paul0403** [comment]
@mehrdad2m note that the kokkos failure did not come from this PR, so nothing to worry about from your end

File: frontend/test/pytest/test_mid_circuit_measurement.py [hunk 1]
```diff
@@ -800,7 +800,7 @@ def cost():
             return [qml.expval(qml.Z(i)) for i in range(qubits)]
 
         result = cost()
-        assert jnp.array(result).shape == (qubits, shots)
+        assert jnp.array(result).shape == (qubits,)
```
Comment (dime10) [inline frontend/test/pytest/test_mid_circuit_measurement.py:803]: That's a weird one, why did we have a `shots` dimension here before 🤔

File: frontend/catalyst/qfunc.py [hunk 2]
```diff
@@ -267,12 +269,23 @@ def wrap_single_shot_qnode(*_):
...
+        if has_mcm:
+            out = parse_native_mid_circuit_measurements(
+                cpy_tape, aux_tapes, results, postselect_mode="pad-invalid-samples"
+            )
+            if len(cpy_tape.measurements) == 1:
+                out = (out,)
+        else:
+            m_count = 0
+            for m in cpy_tape.measurements:
+                is_valid = jnp.array([True] * len(out[m_count]))
```
Comment (dime10) [inline frontend/catalyst/qfunc.py:283]: What does this `is_valid` do?

Comment (dime10) [reply frontend/catalyst/qfunc.py:283]: Oh right, this is about post-selection and filtering invalid samples I presume
```suggestion
                # Without MCMs and postselection, all samples are valid for use in MP computation.
                is_valid = jnp.array([True] * len(out[m_count]))
```

File: frontend/catalyst/qfunc.py [hunk 3]
```diff
@@ -267,12 +269,23 @@ def wrap_single_shot_qnode(*_):
...
+                out[m_count] = gather_non_mcm(
+                    m, out[m_count], is_valid, postselect_mode="pad-invalid-samples"
+                )
+                m_count += 1
```
Comment (dime10) [inline frontend/catalyst/qfunc.py:287]: Slightly cleaner:
```suggestion
            for m_count, m in enumerate(cpy_tape.measurements):
                is_valid = jnp.array([True] * len(out[m_count]))
                out[m_count] = gather_non_mcm(
                    m, out[m_count], is_valid, postselect_mode="pad-invalid-samples"
                )
```

**Pull request merged**